docs/examples/bayesian_optimisation.py

```diff
@@ -0,0 +1,737 @@
+            subkey,
+            lower_bound,
+            upper_bound,
+            num_initial_sample_points=1000,
+        )
+
+        # Evaluate the black-box function at the best point observed so far, and add it to the dataset
+        y_star = six_hump_camel(x_star)
+        print(
+            f"BO Iteration: {i + 1}, Queried Point: {x_star}, Black-Box Function Value: {y_star}"
+        )
+        D = D + gpx.Dataset(X=x_star, y=y_star)
+    bo_experiment_results.append(D)
+
+
+# %% [markdown]
+# We'll also run a random benchmark, whereby we randomly sample from the search space for
+# 20 iterations. This is a useful benchmark to compare the performance of BO against in
+# order to ascertain how much of an advantage BO provides over such a simple approach.
+#
+
+# %%
+random_experiment_results = []
+for i in range(num_experiments):
+    key, subkey = jr.split(key)
+    initial_x = bo_experiment_results[i].X[:5]
+    initial_y = bo_experiment_results[i].y[:5]
+    final_x = jr.uniform(
+        key, shape=(15, 2), dtype=jnp.float64, minval=lower_bound, maxval=upper_bound
+    )
+    final_y = six_hump_camel(final_x)
+    random_x = jnp.concatenate([initial_x, final_x], axis=0)
+    random_y = jnp.concatenate([initial_y, final_y], axis=0)
+    random_experiment_results.append(gpx.Dataset(X=random_x, y=random_y))
+
+
+# %% [markdown]
+# Finally, we'll process the experiment results to find the best observed value of the
+# black-box function at each iteration of the experiments. We'll then take the mean and
+# standard deviation of these values across the 5 experiments.
+
+
+# %%
+def obtain_cumulative_minimum_statistics(
+    experiment_results: List[gpx.Dataset],
+) -> Tuple[Float[Array, "N 1"], Float[Array, "N 1"]]:
+    cumulative_best_observation_results = []
+    for exp_result in experiment_results:
+        observations = exp_result.y
+        cumulative_best_observations = jax.lax.associative_scan(
+            jax.numpy.minimum, observations
+        )
+        cumulative_best_observation_results.append(cumulative_best_observations)
+
+    cumulative_best_observation_results = jnp.array(cumulative_best_observation_results)
+    cumulative_best_observation_mean = jnp.mean(
+        cumulative_best_observation_results, axis=0
+    )
+    cumulative_best_observation_std = jnp.std(
+        cumulative_best_observation_results, axis=0
+    )
+    return cumulative_best_observation_mean, cumulative_best_observation_std
+
+
+bo_cumulative_min_mean, bo_cumulative_min_std = obtain_cumulative_minimum_statistics(
+    bo_experiment_results
+)
+(
+    random_cumulative_min_mean,
+    random_cumulative_min_std,
+) = obtain_cumulative_minimum_statistics(random_experiment_results)
+
+# %% [markdown]
+# Now, when we plot the mean and standard deviation of the best observed value of the
+# black-box function at each iteration, we can see that BO outperforms random sampling,
+# consistently converging to the global minimum of the function.
```

Comment:
we need this to be log(regret) so we can actually see a difference. I would be tempted to run for 15 steps to save a bit of time

Reply:
Have changed to log(regret), and run for 16 steps to show a clear difference whilst saving time.